tls/client-tls-pkcs12.c

```diff
@@ -0,0 +1,269 @@
+    ret = wolfSSL_PKCS12_parse(pkcs12, pass, &pkey, &cert, &ca);
+    if (ret != WOLFSSL_SUCCESS) {
+        printf("Failed to parse the PKCS12 data. ret = %d\n", ret);
+        ret = APP_ERR;
+        goto client_example_end_no_ssl;
+    }
+
+    /* NOTE: these two calls are optional, only required for mutual auth */
+    wolfSSL_CTX_use_PrivateKey(ctx, pkey); /* optional */
+    wolfSSL_CTX_use_certificate(ctx, cert); /* optional */
+
+    /* This step is REQUIRED to validate the peer being connected to */
+    if (ca != NULL) {
+        do {
+            tmp = wolfSSL_sk_X509_pop(ca);
+            if (tmp == NULL && firstLoop) {
+                return ERR_NO_CA_IN_P12;
+            } else if (tmp != NULL) {
+                /* do something interesting with tmp, maybe it's an intermediate
+                 * CA, maybe it's the root CA?
+                 */
+                ret = wolfSSL_X509_get_isCA(tmp);
+                if (ret == 1) {
+                    root_ca_buf = (byte*) wolfSSL_X509_get_der(tmp,
+                                                             &root_ca_buf_size);
+
+                    wolfSSL_CTX_load_verify_buffer(ctx, root_ca_buf,
+                                           root_ca_buf_size, SSL_FILETYPE_ASN1);
+                    wolfSSL_X509_free(tmp);
+                } else {
+                    wolfSSL_X509_free(tmp);
+                }
+            }
+            firstLoop = 0;
+            /* Continue popping until no certs left or target cert is found */
+        } while (tmp != NULL);
+    }
+
+    /* Create a WOLFSSL object */
+    if ((ssl = wolfSSL_new(ctx)) == NULL) {
+        fprintf(stderr, "ERROR: failed to create WOLFSSL object\n");
+        ret = -1;
+        goto client_example_end_no_ssl;
+    }
+
+    /* Attach wolfSSL to the socket */
+    wolfSSL_set_fd(ssl, sockfd);
+
+    /* Connect to wolfSSL on the server side */
+    if (wolfSSL_connect(ssl) != SSL_SUCCESS) {
+        fprintf(stderr, "ERROR: failed to connect to wolfSSL\n");
+        ret =  -1;
+        goto client_example_end;
+    }
+
+    /* Get a message for the server from stdin */
+    printf("Message for server: ");
+    memset(buff, 0, sizeof(buff));
+    fgets(buff, sizeof(buff), stdin);
+    len = strnlen(buff, sizeof(buff));
+
+    /* Send the message to the server */
+    if (wolfSSL_write(ssl, buff, (int) len) != len) {
+        fprintf(stderr, "ERROR: failed to write\n");
+        ret = -1;
+        goto client_example_end;
+    }
+
+    /* Read the server data into our buff array */
+    memset(buff, 0, sizeof(buff));
+    if (wolfSSL_read(ssl, buff, sizeof(buff)-1) == -1) {
+        fprintf(stderr, "ERROR: failed to read\n");
+        ret = -1;
+    }
+
+    /* Print to stdout any data the server sends */
+    printf("Server: %s\n", buff);
+
+client_example_end:
+    /* Cleanup and return */
+    wolfSSL_free(ssl);      /* Free the wolfSSL object                  */
+client_example_end_no_ssl:
```

Comment:
wolfSSL_free should be able to handle a null pointer as an argument. Maybe initialize ssl to NULL and only have one goto. This would reduce complexity for users reading the example code.